disperser/controller/dispatcher.go

```diff
@@ -0,0 +1,432 @@
+		return fmt.Errorf("failed to receive and validate signatures: %w", err)
+	}
+	numPassed, passedQuorums := numBlobsAttestedByQuorum(quorumAttestation.QuorumResults, batchData.Batch.BlobCertificates)
+	if numPassed == 0 {
+		err = d.updateBatchStatus(ctx, batchData.BlobKeys, v2.Failed)
+		if err != nil {
+			d.logger.Error("failed to mark blobs as failed", "err", err)
+		}
+		return fmt.Errorf("no blobs attested by any quorum")
+	}
+
+	nonEmptyQuorums := []core.QuorumID{}
+	for quorumID := range passedQuorums {
+		d.logger.Info("quorum successfully attested", "quorumID", quorumID)
+		nonEmptyQuorums = append(nonEmptyQuorums, quorumID)
+	}
+	aggSig, err := d.aggregator.AggregateSignatures(ctx, d.chainState, uint(batchData.Batch.BatchHeader.ReferenceBlockNumber), quorumAttestation, nonEmptyQuorums)
+	if err != nil {
+		return fmt.Errorf("failed to aggregate signatures: %w", err)
+	}
+	err = d.blobMetadataStore.PutAttestation(ctx, &corev2.Attestation{
+		BatchHeader:      batchData.Batch.BatchHeader,
+		AttestedAt:       uint64(time.Now().UnixNano()),
+		NonSignerPubKeys: aggSig.NonSigners,
+		APKG2:            aggSig.AggPubKey,
+		QuorumAPKs:       aggSig.QuorumAggPubKeys,
+		Sigma:            aggSig.AggSignature,
+		QuorumNumbers:    nonEmptyQuorums,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to put attestation: %w", err)
+	}
+
+	err = d.updateBatchStatus(ctx, batchData.BlobKeys, v2.Certified)
+	if err != nil {
+		return fmt.Errorf("failed to mark blobs as certified: %w", err)
+	}
+
+	return nil
+}
+
+// NewBatch creates a batch of blobs to dispatch
+// Warning: This function is not thread-safe
+func (d *Dispatcher) NewBatch(ctx context.Context, referenceBlockNumber uint64) (*batchData, error) {
+	blobMetadatas, err := d.blobMetadataStore.GetBlobMetadataByStatus(ctx, v2.Encoded, d.lastUpdatedAt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get blob metadata by status: %w", err)
+	}
+
+	if len(blobMetadatas) == 0 {
+		return nil, errNoBlobsToDispatch
+	}
+
+	state, err := d.GetOperatorState(ctx, blobMetadatas, referenceBlockNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get operator state: %w", err)
+	}
+
+	keys := make([]corev2.BlobKey, len(blobMetadatas))
+	for i, metadata := range blobMetadatas {
+		if metadata == nil || metadata.BlobHeader == nil {
+			return nil, fmt.Errorf("invalid blob metadata")
+		}
+		blobKey, err := metadata.BlobHeader.BlobKey()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get blob key: %w", err)
+		}
+		keys[i] = blobKey
+		if metadata.UpdatedAt > d.lastUpdatedAt {
+			d.lastUpdatedAt = metadata.UpdatedAt
+		}
+	}
+
+	certs, _, err := d.blobMetadataStore.GetBlobCertificates(ctx, keys)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get blob certificates: %w", err)
+	}
+
+	if len(certs) != len(keys) {
+		return nil, fmt.Errorf("blob certificates not found for all blob keys")
+	}
+
+	certsMap := make(map[corev2.BlobKey]*corev2.BlobCertificate, len(certs))
```

Comment:
Instead of reordering the certs, would it be simpler to just rederive the `keys` array from the `certs` array?

Reply:
I suppose the current approach forces a fault if there is somehow a redundant cert.

Reply:
Yes, that would be simpler but also true that it enforces cert is present for every key. We'd also need to reorder metadatas (ordering is not a requirement though, it's just for sanity)